crates/locutus-stdlib/src/component_interface.rs

```diff
@@ -0,0 +1,327 @@
+use std::{
+    borrow::{Borrow, Cow},
+    fmt::Display,
+    path::Path,
+};
+
+use blake2::{Blake2s256, Digest};
+use serde::{Deserialize, Serialize};
+use serde_with::serde_as;
+
+use crate::prelude::ContractInstanceId;
+
+const APPLICATION_HASH_SIZE: usize = 32;
+const COMPONENT_HASH_LENGTH: usize = 32;
+
+/// Executable component
+#[derive(Debug, Serialize, Deserialize, Clone)]
+#[serde_as]
+pub struct Component<'a> {
+    #[serde_as(as = "serde_with::Bytes")]
+    #[serde(borrow)]
+    code: Cow<'a, [u8]>,
+    key: ComponentKey,
+}
+
+impl Component<'_> {
+    pub fn key(&self) -> &ComponentKey {
+        &self.key
+    }
+
+    pub fn into_owned(self) -> Component<'static> {
+        Component {
+            key: self.key,
+            code: Cow::from(self.code.into_owned()),
+        }
+    }
+}
+
+impl AsRef<[u8]> for Component<'_> {
+    fn as_ref(&self) -> &[u8] {
+        self.code.borrow()
+    }
+}
+
+impl TryFrom<&Path> for Component<'static> {
+    type Error = std::io::Error;
+    fn try_from(path: &Path) -> Result<Self, Self::Error> {
+        let code = Cow::from(std::fs::read(path)?);
+        let key = ComponentKey::new(code.borrow());
+        Ok(Self { code, key })
+    }
+}
+
+/// Type of errors during interaction with a component.
+#[derive(Debug, thiserror::Error, Serialize, Deserialize)]
+pub enum ComponentError {
+    #[error("de/serialization error: {0}")]
+    Deser(String),
+    #[error("{0}")]
+    Other(String),
+}
+
+#[serde_as]
+#[derive(Clone, Serialize, Deserialize, Debug, PartialEq, Eq)]
+pub struct SecretsId {
+    #[serde_as(as = "serde_with::Bytes")]
+    key: Vec<u8>,
+    #[serde_as(as = "[_; 32]")]
+    hash: [u8; 32],
+}
+
+impl SecretsId {
+    pub fn new(key: Vec<u8>) -> Self {
+        let mut hasher = Blake2s256::new();
+        hasher.update(&key);
+        let hashed = hasher.finalize();
+        let mut hash = [0; 32];
+        hash.copy_from_slice(&hashed);
+        Self { key, hash }
+    }
+
+    pub fn encode(&self) -> String {
+        bs58::encode(self.hash)
+            .with_alphabet(bs58::Alphabet::BITCOIN)
+            .into_string()
+    }
+}
+
+pub trait ComponentInterface {
+    /// Process inbound message, producing zero or more outbound messages in response
+    /// Note that all state for the component must be stored using the secret mechanism.
+    fn process(messages: InboundComponentMsg) -> Result<Vec<OutboundComponentMsg>, ComponentError>;
+}
+
+#[serde_as]
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Serialize, Deserialize)]
+pub struct ComponentKey(#[serde_as(as = "[_; COMPONENT_HASH_LENGTH]")] [u8; COMPONENT_HASH_LENGTH]);
+
+impl ComponentKey {
+    pub fn new(wasm_code: &[u8]) -> Self {
+        let mut hasher = Blake2s256::new();
```

Comment:
Should components have a parameters `[u8]` similar to contract parameters, to make them easier to configure/customize without having to recompile the wasm?

Reply:
maybe, but I thought for now we decided we wouldn't need them. We can add parameters if needed in the future.